src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -332,18 +341,53 @@ await CallWithRequestTracing(
+                                watchedFeatureFlagCollections = await LoadSelected(data, _options.FeatureFlagSelectors, client, cancellationToken).ConfigureAwait(false);
                                 watchedSettings = await LoadKeyValuesRegisteredForRefresh(client, data, cancellationToken).ConfigureAwait(false);
-                                watchedSettings = UpdateWatchedKeyValueCollections(watchedSettings, data);
                                 logInfoBuilder.AppendLine(LogHelper.BuildConfigurationUpdatedMessage());
                                 return;
                             }
+                            else
+                            {
+                                // Get key value collection changes if RegisterAll was called
+                                watchedSelectedKvCollections = _watchedSelectedKvCollections;
 
-                            changedKeyValuesCollection = await GetRefreshedKeyValueCollections(refreshableMultiKeyWatchers, client, logDebugBuilder, logInfoBuilder, endpoint, cancellationToken).ConfigureAwait(false);
+                                if (_options.RefreshOptions.RegisterAllEnabled && registerAllIsRefreshable)
+                                {
+                                    bool selectedKvCollectionsChanged = await UpdateWatchedCollections(_options.KeyValueSelectors, watchedSelectedKvCollections, client, cancellationToken).ConfigureAwait(false);
 
-                            if (!changedKeyValuesCollection.Any())
-                            {
-                                logDebugBuilder.AppendLine(LogHelper.BuildFeatureFlagsUnchangedMessage(endpoint.ToString()));
+                                    if (!selectedKvCollectionsChanged)
+                                    {
+                                        logDebugBuilder.AppendLine(LogHelper.BuildSelectedKeyValueCollectionsUnchangedMessage(endpoint.ToString()));
+                                    }
+                                    else
+                                    {
+                                        keyValueChanges.AddRange(await GetRefreshedCollections(_options.KeyValueSelectors, _mappedData, client, cancellationToken).ConfigureAwait(false));
+
+                                        logInfoBuilder.Append(LogHelper.BuildSelectedKeyValueCollectionsUpdatedMessage());
+                                    }
+                                }
+
+                                // Get feature flag changes
+                                watchedFeatureFlagCollections = _watchedFeatureFlagCollections;
+
+                                bool featureFlagCollectionsChanged = await UpdateWatchedCollections(
+                                    refreshableFeatureFlagWatchers.Select(watcher => new KeyValueSelector { KeyFilter = watcher.Key, LabelFilter = watcher.Label }),
```

Comment:
The reason why `watchedFeatureFlagCollections` cannot be used here is that we maintained a refresh timer for each feature flag selector. But for registerAll, you used `_registerAllNext`. 


Here we treat key value collection differently from feature flag collection. Actually, I prefer the way you used for key value collection. A shared refresh timer is enough. No need to maintain a unqiue refresh timer for each feature flag selector. It just makes our code looks so complicated.
IMO, for collection monitoring based refresh, a refresh timer for all key value collections and a refresh timer for all feature flags are enough.


Even more agressively, I think maintaining a unique refresh timer for each KeyValueWatcher is over engineering. In the future, I think we should remove it.

Reply:
> No need to maintain a unqiue refresh timer for each feature flag selector.

.NET provider supports different refresh intervals for different feature flag selectors. The reason is the same why we support different refresh intervals for sentinel keys - its because options can be stacked.

`ConfigureRefresh` can be invoked multiple times with different sentinel keys and their respective refresh interval. Similarly, `UseFeatureFlags` can be invoked multiple times with different FF selectors and their respective refresh intervals. Therefore, we need to support different refresh intervals for sentinel keys and feature flags.

For selected kv collections, we will support only one refresh interval.

Reply:
I see. In JS provider, user cannot call `UseFeatureFlags` for multiple times. All feature flag selectors are listed at the same time. So they will use shared refresh interval.

Now, I understand the difference of .NET provider. Maintaining different refresh intervals for different ff selectors makes sense to me.